volatility3/framework/layers/qemu.py

```diff
@@ -192,6 +235,63 @@ def _load_segments(self):
             else:
                 raise exceptions.LayerException(self._name, f'QEMU unknown section encountered: {section_byte}')
 
+        # If no architecture has been set, try to determine it using fallback mechanisms
+        if not self._architecture:
+            self._architecture = self._fallback_determine_architecture()
+            if self._architecture is None:
+                vollog.log(constants.LOGLEVEL_VV, f"QEVM architecture could not be determined")
+
+        # Once all segments have been read, determine the PCI hole if any
+        for regex in self.pci_hole_table:
+            if regex.match(self._architecture):
+                self._pci_hole_minimum, self._pci_hole_start, self._pci_hole_end = self.pci_hole_table[regex]
+                if self.maximum_address < self._pci_hole_minimum:
+                    # The PCI hole isn't present because we're below the minimum value
+                    self._pci_hole_start, self._pci_hole_end = 0, 0
+                vollog.log(constants.LOGLEVEL_VVVV, f"QEVM architecture detected as: {self._architecture}")
+                break
+        else:
+            vollog.log(constants.LOGLEVEL_VVVV, f"QEVM unknown architecture found: {self._architecture}")
+
+    def _fallback_determine_architecture(self) -> str:
+        architecture_pattern = rb'pc-(i440fx|q35)-([0-9]{1,2}.[0-9]{1,2}(?:.[0-9]{1,2})?)'
+        old_suffix = "-1.0"
+        base_layer = self.context.layers[self._base_layer]
+
+        vollog.log(constants.LOGLEVEL_VVVV, "QEVM fallback architecture detection used")
+
+        res = scanners.RegExScanner(architecture_pattern)
+        for offset in base_layer.scan(context = self.context, scanner = res):
+            line = base_layer.read(offset, 64)
+            regex_results = re.search(architecture_pattern, line)
+            architecture = "pc-" + regex_results.groups()[0].decode() + old_suffix
```

Comment:
Don't use the suffix here, since the regex contains the version number. Use `f"pc-{regex_results.groups()[0].decode()-{regex_results.groups()[1].decode()}"` instead.

Reply:
Yeah, that feels really ugly, I'd prefer to not using the grouping and just return the whole match, except I was a little too lazy to figure out how to do it when making the changes.  So I'll turn this into the entire match and hopefully that'll satisfy both of us...  5;)